pkg/controllers/sandbox/unsandboxsubcluster_reconciler.go

```diff
@@ -0,0 +1,267 @@
+			return ctrl.Result{}, nil
+		}
+		return r.unsandboxSubclusters(ctx)
+	}
+	return ctrl.Result{}, nil
+}
+
+// reconcileSandboxStatus will update sandbox status for the subclusters that are already unsandboxed
+func (r *UnsandboxSubclusterReconciler) reconcileSandboxStatus(ctx context.Context) error {
+	scSbInStatus := make(map[string]string)
+	for _, sb := range r.Vdb.Status.Sandboxes {
+		for _, sc := range sb.Subclusters {
+			scSbInStatus[sc] = sb.Name
+		}
+	}
+	sbScMap := r.PFacts.FindUnsandboxedSubclustersStillInSandboxStatus(scSbInStatus)
+	for sb, scs := range sbScMap {
+		err := r.updateSandboxStatus(ctx, sb, scs)
+		if err != nil {
+			r.Log.Error(err, "failed to update sandbox status", "sandbox", sb, "new subclusters", scs)
+			return err
+		}
+	}
+
+	return nil
+}
+
+// reconcileSandboxConfigMap will update/delete sandbox config map if it expires, this function will return
+// an error and a boolean to indicate if sandbox config map is deleted
+func (r *UnsandboxSubclusterReconciler) reconcileSandboxConfigMap(ctx context.Context) (error, bool) {
+	sbName := r.ConfigMap.Data[vapi.SandboxNameKey]
+	cmName := r.ConfigMap.Name
+	sb := r.Vdb.GetSandboxStatus(sbName)
+	// if the sandbox doesn't have any subclusters, we delete the config map
+	if sb == nil || len(sb.Subclusters) == 0 {
+		err := r.Client.Delete(ctx, r.ConfigMap)
+		if err != nil {
+			r.Log.Error(err, "failed to delete expired sandbox config map", "configMapName", cmName)
+			return err, false
+		}
+		r.Log.Info("deleted expired sandbox config map", "configMapName", cmName)
+		return nil, true
+	}
+	if r.ConfigMap.Annotations[vmeta.SandboxControllerUnsandboxTriggerID] != "" {
+		unsandboxSbScMap := r.Vdb.GenSandboxSubclusterMapForUnsandbox()
+		_, found := unsandboxSbScMap[sbName]
+		// if the subclusters in the sandbox does not need to be unsandboxed, we remove
+		// unsandbox trigger ID from the config map
+		if !found {
+			delete(r.ConfigMap.Annotations, vmeta.SandboxControllerUnsandboxTriggerID)
+			err := r.Client.Update(ctx, r.ConfigMap)
+			if err != nil {
+				r.Log.Error(err, "failed to remove unsandbox trigger ID from sandbox config map", "configMapName", cmName)
+				return err, false
+			}
+			r.Log.Info("Successfully removed unsandbox trigger ID from sandbox config map", "configMapName", cmName)
+		}
+	}
+
+	return nil, false
+}
+
+// unsandboxSubclusters will unsandbox subclusters inside the sandbox from sandbox config map
+func (r *UnsandboxSubclusterReconciler) unsandboxSubclusters(ctx context.Context) (ctrl.Result, error) {
+	// find an initiator to call vclusterOps
+	initiatorIP, ok := r.PFacts.FindFirstPrimaryUpPodIP()
+	if ok {
+		r.InitiatorIP = initiatorIP
+	} else {
+		r.Log.Info("Requeue because there are no UP nodes in main cluster to execute unsandbox operation")
+		return ctrl.Result{Requeue: true}, nil
+	}
+
+	err := r.executeUnsandboxCommand(ctx)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+	r.PFacts.Invalidate()
```

Comment:
`r.PFacts` is just a local copy since it collects for the main cluster. We should invalidate the pfacts that was passed into the contructor. As that is the one that has done collection for the sandbox.

Reply:
We only made a copy of pfacts that was passed into the constructor. Is there a need to invalidate that one? I think I can simply not invalidate r.PFacts here since unsandbox reconciler will collect podFacts every time it runs.